### ISSUE TITLE
SHACL syntax validation fails for sh:or with "Value does not have shape shsh:ListShape"

### ISSUE THREAD

**Issue by jaehnigen-kp** [BrickSchema/Brick#656]
After having issues with trying to import `brick.ttl` into engines like Stardog I ran `brick.ttl` (version 1.4.1) through the SHACL shape validator at https://www.itb.ec.europa.eu/shacl/shacl/upload (against the "Core W3C syntax rules"). This reports 160 errors with the message:

```
Value does not have shape shsh:ListShape
Test: [Value] - [https://brickschema.org/schema/BrickShape#NumericValue]
Location: [Result path] - [http://www.w3.org/ns/shacl#or
``` 

This is also an error that the Stardog Studio Text Editor reports and that seems to break the Stardog import.

Further research led me to https://www.w3.org/TR/shacl/#OrConstraintComponent which AFAICT requires a list syntax for the values.

Can you confirm my assumption that this is a bug that should be fixed?

**Comment by gtfierro** [comment]
Hi @jaehnigen-kp !

Here's what the Turtle serialization of `bsh:NumericValue` looks like in the latest distribution of Brick:

```ttl
bsh:NumericValue a rdf:List ;
    rdf:first [ sh:datatype xsd:float ] ;
    rdf:rest ( [ sh:datatype xsd:decimal ] [ sh:datatype xsd:integer ] [ sh:datatype xsd:double ] ) .
```

I believe this is a valid List per the SHACL list definition (https://www.w3.org/TR/shacl/#dfn-shacl-list). We haven't had any issues with this construct in pySHACL or Topquadrant's SHACL.

I found a definition of `shsh:ListShape` [here](https://github.com/RDFLib/pySHACL/blob/master/test/resources/sht_tests/core/complex/shacl-shacl-data-shapes.ttl#L21-L31) and there's nothing obviously wrong to me. I'll poke around and see if I can find why this is raising an error

**Comment by jaehnigen-kp** [comment]
Hey @gtfierro, for a first step I assumed that the problem is with the `sh:or` value (should be SHACL list according to https://www.w3.org/TR/shacl/#OrConstraintComponent) not being enclosed in brackets. Therefore I changed all (60) occurances of `sh:or bsh:NumericValue ;` to `sh:or ( bsh:NumericValue ) ;` which gave me 60 validation errors less. But I still got 90 errors with the same error message as above. So that seems to not have been the final issue. 

What's your recommendation about these kinds of validation errors in general? There seem to be some contradicting validators out there. Do you know about that specific validator? Is there recommended/"definite" one?

**Comment by steveraysteveray** [comment]
Just jumping in, but I think sh:or expects a list of SHACL shapes. Try 
```
sh:or ( [bsh:NumericValue] )
```

**Comment by gtfierro** [comment]
If I have `sh:or bsh:NumericValue` and `bsh:NumericValue` is defined as I did above (https://github.com/BrickSchema/Brick/issues/656#issuecomment-2334366394), doesn't that fulfill the necessary definition?

**Comment by steveraysteveray** [comment]
I see what you mean...hmm.

**Comment by gtfierro** [comment]
@jaehnigen-kp I think this might be a bug in Stardog. Chances are (?) they expect the list (the object of `sh:or`) to be a blank node. However, we re-use the list here in several places so it's represented as a named node. It is possible we could refactor this to use a Nodeshape to contain the `sh:or`, and just compose that with the shapes that need it. However,  without anything more specific in terms of logs, it is hard to know what is going wrong. The other SHACL processors we use have not complained about this construct.

We have had success using:
- the TopQuadrant implementation (pretty fast, but you have to interact in Java or over CLI): https://github.com/TopQuadrant/shacl
- pySHACL (slower, but in Python): https://github.com/RDFLib/pySHACL
- there are others listed here: https://w3c.github.io/data-shapes/data-shapes-test-suite/#implementation-reports

Be aware that you will want to choose implementations that support SHACL-AF: https://www.w3.org/TR/shacl-af/

Let me know if you have any questions!

**Comment by jaehnigen-kp** [comment]
@gtfierro Please be aware that I was primarily refering to the validator at https://www.itb.ec.europa.eu/shacl/shacl/upload. Have you tried that yourself? The validation results of Stardog were that confusing that I at first wanted to find some easy-to-access validator and thus ended up at the one linked above. I'll get back when I got some more feedback that is worth following up on.

**Comment by gtfierro** [comment]
Sorry for my confusing reply! I have tried that site, and I have no idea why it is reporting the results it is. As far as I can tell, our list is correctly formatted using SHACL. Do you get any better report results out of Stardog?

**Comment by jaehnigen-kp** [comment]
> Do you get any better report results out of Stardog?

Unfortunately not, even worse. For now I only have access to Stardog Cloud and that provides only an online editor view for the TTL with some inline error display. I'll try to get some feedback from there as well.

**Comment by gtfierro** [comment]
Hi @costas80 

Here is a short script which downloads the latest Brick and runs it through the TopQuadrant validation. I have not been able to replicate the SHACL error mentioned by @jaehnigen-kp except on the Interop TestBed site. There are a couple errors here but they mostly have to do with missing imports which shouldn't be relevant to the error. 

```bash
wget https://repo1.maven.org/maven2/org/topbraid/shacl/1.4.3/shacl-1.4.3-bin.zip
unzip shacl-1.4.3-bin.zip
pushd shacl-1.4.3
export SHACLROOT=`pwd`/bin
export PATH=$SHACLROOT:$PATH
chmod +x $SHACLROOT/*.sh
wget https://brickschema.org/schema/1.4.1/Brick.ttl
shaclvalidate.sh -datafile Brick.ttl
```

**Comment by costas80** [comment]
Hi @gtfierro 

Having looked into this I believe there is a misunderstanding here. When you say that different SHACL processors don't complain about the Brick shapes, this is when you are using the Brick shapes to validate a graph. This is also what you did in the TopQuadrant PoC that you shared:
```
shaclvalidate.sh -datafile Brick.ttl
```
What you did here is the same as using our generic RDF validator at https://www.itb.ec.europa.eu/shacl/any/upload, if you supply both for the content to validate and the shapes to use the Brick shapes. The validation report when comparing the TopQuadrant CLI result and our RDF validator are identical which is not surprising given that internally our validator also uses the TopQuadrant (albeit through Java).

This brings me to the initial issue raised by @jaehnigen-kp, who said he used the Brick shapes with our SHACL shape validator at https://www.itb.ec.europa.eu/shacl/shacl/upload with the "Core W3C syntax rules". This validator does not validate a graph using the Brick shapes; it rather considers the Brick shapes as the input to validate and validates them against the shapes of the configured profiles. The "Core W3C syntax rules" mentioned are the ones listed in https://www.w3.org/TR/shacl/#shacl-shacl in Section C, "SHACL Shapes to Validate Shapes Graphs". If you take the listed shapes from there, save them in say `core_shapes.ttl` and use them to validate the Brick shapes (`shaclvalidate.sh -datafile Brick.ttl -shapesfile core_shapes.ttl`), you will see 160 errors exactly as @jaehnigen-kp mentioned, and matching what our SHACL shape validator reports:
```
  sh:result    [ rdf:type                      sh:ValidationResult;
                 sh:focusNode                  [] ;
                 sh:resultMessage              "Value does not have shape shsh:ListShape";
                 sh:resultPath                 sh:or;
                 sh:resultSeverity             sh:Violation;
                 sh:sourceConstraintComponent  sh:NodeConstraintComponent;
                 sh:sourceShape                _:b0;
                 sh:value                      bsh:NumericValue
               ];
```
So, to summarise, the issue at hand is that the Brick shapes fail validation against the published core shapes at https://www.w3.org/TR/shacl/#shacl-shacl (irrespective of our validator). You could choose to address this or not in the Brick shapes but at least you should investigate further and see if a fix is needed.

**Comment by gtfierro** [comment]
I'm fairly confident this is due to a circular import of Brick through the REC ontology. I will explore a few fixes and see what we can land in the next Brick patch release (1.4.2)

**Comment by jaehnigen-kp** [comment]
@gtfierro This version fully validates against the Core W3C rules in @costas80 validator: 
![grafik](https://github.com/user-attachments/assets/127765b9-955a-4c7a-a6f4-4a97c540c5d6)
I'll see if that also shows improvements in Stardogs validation and be back in a separate issue if there is proper evidence to follow up. Thank you both for your great support, I learned quite a bit in the process!
